pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -0,0 +1,220 @@
+func (r *RegURL) String() string {
+	return r.URL.Host + r.URL.Path
+}
+
+// newRegURL parses the input string, performs some sanity checks and returns
+// a url.URL.  The input must be a valid URI with an "http" or "https" scheme,
+// a specified host and an empty URI user.  Otherwise, an error is returned.
+func newRegURL(input string) (url.URL, error) {
+	uri, err := url.Parse(input)
+	if err != nil {
+		return url.URL{}, fmt.Errorf("error parsing URL %s: %v", input, err)
+	}
+
+	// only https and http are valid URI schemes
+	if uri.Scheme == "" {
+		return url.URL{}, fmt.Errorf("unspecified URI scheme: %s", input)
+	}
+	if uri.Scheme != "https" && uri.Scheme != "http" {
+		return url.URL{}, fmt.Errorf("unsupported URI scheme: %s", uri.Scheme, input)
+	}
+
+	// a host must be specified
+	if uri.Host == "" {
+		return url.URL{}, fmt.Errorf("unspecified URI host: %s", input)
+	}
+
+	// URI user must be empty
+	if uri.User != nil {
+		// strip password for security reasons
+		uri.User = url.UserPassword(uri.User.Username(), "xxxxxx")
+		return url.URL{}, fmt.Errorf("unsupported username/password: %q", uri)
+	}
+
+	return *uri, nil
+}
+
+// UnmashalText turns text into a RegURL, and is a way to unmarshal a string
+// into a url.URL.
+func (r *RegURL) UnmarshalText(text []byte) (err error) {
+	r.URL, err = newRegURL(string(text))
+	return err
+}
+
+// Registry is a generalized type representing a registry host and a registry
+// mirror, whereas the mirror cannot further be mirrored.
+type Registry struct {
+	URL      RegURL     `toml:"url"`
+	Blocked  bool       `toml:"blocked"`
+	Insecure bool       `toml:"insecure"`
+	Search   bool       `toml:"unqualified-search"`
+	Prefix   string     `toml:"prefix"`
+	Mirrors  []Registry `toml:"mirror"`
+}
+
+// String returns host:port/path of the Registry's URL.
+func (r *Registry) String() string {
+	return r.URL.String()
+}
+
+// tomlConfig is the data type used to marshal and unmarshal the toml config.
+type tomlConfig struct {
+	Registries []Registry `toml:"registry"`
+}
+
+// getUnqualifiedSearchRegistries returns all registries that are configured
+// for unqualified image search (i.e., with Registry.Search == true).
+func getUnqualifiedSearchRegistries(registries []Registry) []Registry {
+	unqualified := []Registry{}
+	for _, reg := range registries {
+		if reg.Search {
+			unqualified = append(unqualified, reg)
+		}
+	}
+	return unqualified
+}
+
+// FindRegistries returns all registries matching the specified image reference.
+// In case of an unqualified image reference, all registries that are configured
+// for unqualified image search are returned.  In case of a fully qualified
+// image reference, the registry with the longest prefix is returned.  An empty
+// array is returned in case no matching registry is found.
+func FindRegistries(imgRef reference.Reference, registries []Registry) ([]Registry, error) {
+	domain := reference.Domain(imgRef.(reference.Named))
```

Comment:
If the function crashes with a non-`reference.Named`, why not make the input parameter a `reference.Named` in the first place?

---

I’m not sure this is going to actually work, though; `c/image/docker/reference.Reference` is not the Fedora `/usr/bin/docker`-like `docker/docker/reference` package which implements unqualified names as a first-class concept, its the upstream-like one where a reference always implicitly includes a domain. `ParseNormalizedNamed` will add a `docker.io` host name; `ParseNamed` will refuse an unqualified name.

`reference.Parse` does make it possible to create a `domain`-less value, but that value would be dangerous to pass to any other user in c/image which expects normalization to happen. And it incorrectly parses `notlocalhost/busybox` as domain=`notlocalhost`, path=`busybox`, instead of domain=unknown, path=`notlocalhost/busybox`.

[The search semantics is _really problematic_ for security: with it, and a `pull fedora` command, the user never knows what image will be downloaded.]

If this needs to exist at all (which, sadly, seems to be the case), the `imgRef` should probably just be a `string`; `reference.Named` would only be created by the result of combining the domain and the input string.

---

And maybe the public API should not combine the unqualified/qualified cases at all; AFAICT the caller needs to treat them differently anyway (for unqualified, combine the search registry and the input; for qualified, use the input as specified, and only use the flags), so the caller can just as well call two different public API endpoints here (`GetUnqualifiedSearchRegistries`/`GetRegistryForReference` or something like that); then, the qualified case could return a `*Registry` instead of an array, and the caller would not need to worry about what happens when there is >1 element.